router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
```

Comment:
> But we aren't doing anything (e.g. reading the request body) to trigger that `CloseNotifier`.

Not sure I follow. We don't have to "trigger" the `CloseNotifier`; that could happen anytime after we call `CloseNotify()`: https://github.com/golang/go/blob/928c83ff2cd13e5973827b2abde43b6dc901ec85/src/net/http/server.go#L173-L180

Anyway, the current iteration of this does seem the most correct of the ones we've had so far. Admittedly, though, it does take a lot of digging to really understand why it works.